### ISSUE TITLE
Share image texture in memory between platform and Flutter

### ISSUE THREAD

**Issue by xster** [flutter/flutter#59410]
In add-to-app, the same image may appear both on the platform side and in Flutter. We've made https://pub.dev/packages/ios_platform_images to help share the image asset in storage but we should share the image in memory as well. 

This was accomplished via https://medium.com/hackernoon/rendering-external-texture-an-flutter-optimization-by-alibaba-c5ed143af747 by Alibaba. We should upstream that solution in the Flutter engine. They also sent us a PR (never merged) https://github.com/flutter/engine/commit/dfe3d986c96a29ea40a4c9adc3eb9e811c28a64e. 

cc @gaaclarke @chinmaygarde

**Comment by chinmaygarde** [comment]
The pixel buffer interface is intentional and allows plugins that provide external textures for inline composition to provide these textures in a manner that is agnostic to client rendering API used by Flutter. This forms the basis for the switch from OpenGL to Metal seamlessly. That patch punches a hole in the API and makes assumptions about the Flutter using OpenGL internally. Assumptions which will now break because Flutter can use Metal on newer versions of iOS. For this reason, that patch cannot be upstreamed.

The pixel buffer interface does not prevent applications from this workflow however. The application can obtain the pixel buffer from the platform, realize a texture from it using an application side texture cache with its own context, perform mutations on the texture and pass the pixel buffer reference to the engine for composition. This should maintain the rendering API agnosticism while allowing for the use-case described in the article.

It may be possible to make this workflow more ergonomic but upstreaming that patch is no viable. For this reason, I am treating this as a feature request instead of a request to land the patch as-is.

**Comment by gaaclarke** [comment]
Seems like they could make a plugin that does what they want then.  We just have to verify there is a way for plugins to query the rendering api though so it can change its logic depending on metal or opengl.

**Comment by xster** [comment]
Yes agreed. The intent is to capture the spirit behind the PR and make it more accessible for users who don't patch their engine rather than landing that PR as is.

**Comment by chinmaygarde** [comment]
> We just have to verify there is a way for plugins to query the rendering api though so it can change its logic depending on metal or opengl.

Rendering API agnosticism means that developers shouldn't have to know of care what Flutter is using. They can continue to use OpenGL to mutate their textures before giving it to Flutter for rendering with Metal. [The Wiki entry on texture interop](https://github.com/flutter/flutter/wiki/Metal-on-iOS-FAQ#if-my-application-uses-opengl-to-provide-textures-for-inline-composition-in-a-flutter-application-will-this-work-with-the-metal-backend) may be a useful for folks trying to figure out how to achieve this.

**Comment by gaaclarke** [comment]
That's cool, here is the jist:
```objc
NSDictionary* cvBufferProperties = @{
    (__bridge NSString*)kCVPixelBufferOpenGLCompatibilityKey : @YES,
    (__bridge NSString*)kCVPixelBufferMetalCompatibilityKey : @YES,
};
CVReturn cvret = CVPixelBufferCreate(kCFAllocatorDefault,
                        size.width, size.height,
                        _formatInfo->cvPixelFormat,
                        (__bridge CFDictionaryRef)cvBufferProperties,
                        &_CVPixelBuffer);
```

That compatibility feature might get yanked though with OpenGL so still probably ideal to switch your plugin logic on the rendering API.

**Comment by chinmaygarde** [comment]
> That compatibility feature might get yanked though with OpenGL so still probably ideal to switch your plugin logic on the rendering API.

OpenGL is already deprecated. If it is removed completely, the plugin/application won't be able to work with the texture using OpenGL in the first place. At that point, the plugin has nothing to switch between. So this is not a concern.

**Comment by pdc-tstream** [comment]
I just wanted to bump this (hopefully that's not frowned upon) -  it's extremely important to us.

I agree that it's great to have an agnostic path if you want to keep that.  I get it.

For efficiency I believe it's also reasonable to have an API to share textures natively.  Perfectly fine if that means more work for the plugin developer who wants that speed.  If things change, then it's up to them to keep up to date.  No problem - you want the efficient path, you work for it.

Without the ability to pass a texture directly to Flutter (no intermediate PixelBuffer / no round trip to CPU), we are stuck on the sidelines wishing we could make the jump and start using this wonderful platform.

This is important to us as we do a ton of native manipulation of textures, and we stream the results to disk/network and want to see all of that, efficiently, in app.   The round trip, unfortunately, prohibits us from moving to Flutter.  We don't want to hack up Flutter as the guys at Alibaba did because maintenance would be overly burdensome and risky.

Anyway, I just wanted to let you guys know that this is very important to some developers as without it there's a huge stumbling block to move to this platform (which we are desperate to do!)

Anyway, keep up the great work.

**Comment by bazmichael** [comment]
Any updates?..

**Comment by renanyoy** [comment]
any update

**Comment by nikage** [comment]
Thanks @gaaclarke !
This helped me to fix the error.
> NSDictionary* cvBufferProperties = @{
>     (__bridge NSString*)kCVPixelBufferOpenGLCompatibilityKey : @YES,
>     (__bridge NSString*)kCVPixelBufferMetalCompatibilityKey : @YES,
> };
> CVReturn cvret = CVPixelBufferCreate(kCFAllocatorDefault,
>                         size.width, size.height,
>                         _formatInfo->cvPixelFormat,
>                         (__bridge CFDictionaryRef)cvBufferProperties,
>                         &_CVPixelBuffer);

Here is my swift version, in case someone might find it helpful:
```swift
let pixelBufferAttributes: CFDictionary? = [
        kCVPixelBufferOpenGLCompatibilityKey: true,
        kCVPixelBufferMetalCompatibilityKey: true
        ] as CFDictionary

let pixelBufferStatus = CVPixelBufferCreate(
    kCFAllocatorDefault,
    width,
    height,
    kCVPixelFormatType_420YpCbCr8BiPlanarFullRange,
    pixelBufferAttributes,
    &pixelBuffer
)
```

**Comment by renanyoy** [comment]
you can use https://api.flutter.dev/ios-embedder/protocol_flutter_texture_registry-p.html  easy interface to share images between the native/platform and flutter

**Comment by nikage** [comment]
> you can use https://api.flutter.dev/ios-embedder/protocol_flutter_texture_registry-p.html easy interface to share images between the native/platform and flutter , this interface can be retrieved from registrar.textures

Thanks, @renanyoy! 
Actually I'm using FlutterTexture to display video frames using gstreamer. I just got the same error as above while passing `CVPixelBuffer` into flutter texture. 

PS. I'm kinda new to video/image processing and the stuff, so I'm not sure this is the best way to do it.

**Comment by renanyoy** [comment]
@nikage you can look at camera plugin that use it https://github.com/flutter/packages/tree/main/packages/camera
